### PR TITLE
[Feature](bangc-ops): add pad op

### PR DESCRIPTION
**Pull request by gaotianyi544**
## 1. Motivation

将pad算子以二进制形式合入mluops。

## 2. Modification

新增pad算子

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff3
    - [ ] mlu diff3 == 0

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [x] MLU370
  - [x] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

image

Platform：MLU590
image

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/pad/pad.cpp [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include <algorithm>
+#include <string>
+#include "pad.h"
+
```
Comment (PetrelYy) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/pad/pad.cpp:15]:
```suggestion
#include "pad.h"

#include <algorithm>
#include <string>
```

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -3047,6 +3047,94 @@ mluOpAddN(mluOpHandle_t handle,
...
+ *    dimension \p k, \p paddings[k, 0] and \p paddings[k, 1] specify the padding size before and after the input data
+ *    respectively.
+ *    - \p output[k] = \p paddings[k, 0] + \p input[k] + \p paddings[k, 1]. \p k represents each dimension of \b input.
+ *    \p output[k] represents the length of the dimension \p k in the output tensor. \p input[k] represents the length
+ *    of the dimension \p k in the input tensor. \p paddings[k, 0] and \p paddings[k, 1] represent the length to be
+ *    padded before and after the dimension \p k in the input tensor respectively.
+ *    - \p paddings[k, 0] and \p paddings[k, 1] >= \p -input[k]. \p k represents each dimension of \b input.
+ *    \p input[k] represents the length of the dimension \p k in the input tensor.
+ *
+ *  @note
```
Comment (PetrelYy) [inline bangc-ops/mlu_op.h:3099]:
```suggestion
 * @par Note
```

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -3047,6 +3047,94 @@ mluOpAddN(mluOpHandle_t handle,
...
+ *  - None.
+ *
+ *  @par Requirements
+ *  - None.
+ *
+ *  @par Example
+ *  - The example of pad is as follows:
+    @verbatim
+    For example below, if dimensions of input == 2, then the shape of paddings should be [n, 2] == [2, 2], where n holds the
```
Comment (PetrelYy) [inline bangc-ops/mlu_op.h:3108]: 缩进和格式参考文档规范

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/pad/pad.cpp (hunk 1 above)
Comment (gaotianyi544) [reply bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/pad/pad.cpp]: 已修改

File: bangc-ops/mlu_op.h (hunk 2 above)
Comment (gaotianyi544) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 3 above)
Comment (gaotianyi544) [reply bangc-ops/mlu_op.h]: 已修改

**Review by Unireverse — APPROVED** [review]

**Review by liying-1997 — APPROVED** [review]

**Review by PetrelYy — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -3047,6 +3047,91 @@ mluOpAddN(mluOpHandle_t handle,
           const mluOpTensorDescriptor_t output_desc,
           void *output);
 
+// Group:Pad
+/*!
+ * @brief Pads the input tensor using \b padding_value based on \b paddings that specifies how the tensor is
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:3052]: 这句描述再斟酌一下？如果是有cropped的情况，还能用Pads做谓语嘛？

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -3047,6 +3047,91 @@ mluOpAddN(mluOpHandle_t handle,
...
+ * padded or cropped. This operation is usually used in ResNet and Transformer networks.
+ *
+ * @param[in] handle
+ * Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues in the pad operation. For
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:3056]:
```suggestion
 * Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues in the padding operation. For
```

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -3047,6 +3047,91 @@ mluOpAddN(mluOpHandle_t handle,
...
+ * detailed information, see ::mluOpHandle_t.
+ * @param[in] input_desc
+ * The descriptor of the input tensor. For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] input
+ * Pointer to the MLU memory that stores the input tensor.
+ * @param[in] paddings
+ * A host pointer to the \b paddings data that holds the padding size to be added for each dimension
+ * of \b input. Positive and negative padding values represent the padding size and the cropping size, respectively.
+ * @param[in] padding_value
+ * A host pointer to the \b padding_value that holds the constant value to be added.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:3066]: 通常参数说明不会用它本身去解释自己。。。要么换个说法？

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -3047,6 +3047,91 @@ mluOpAddN(mluOpHandle_t handle,
...
+    padding_value: 0
+    output: [[0, 0, 0], [4, 0, 0], [3, 0, 0], [0, 0, 0]]
+   @endverbatim
+ *
+ * @par Reference
+ * - https://www.tensorflow.org/api_docs/python/tf/pad
+ * - https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/array_ops.py
+ * - https://pytorch.org/docs/stable/nn.functional.html?highlight=pad#torch.nn.functional.pad
+ */
+
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:3125]: 删掉多余空行吧~可能会影响pdf生成效果

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -3047,6 +3047,91 @@ mluOpAddN(mluOpHandle_t handle,
...
+ *  supported data types are as follows:
+ *   - input tensor: uint8, int8, uint16, int16, uint32, int32, uint64, int64, bool, half, float.
+ *   - paddings: int32.
+ *   - padding_value: uint8, int8, uint16, int16, uint32, int32, uint64, int64, bool, half, float.
+ *   - output tensor: uint8, int8, uint16, int16, uint32, int32, uint64, int64, bool, half, float.
+ *
+ * @par Scale Limitation
+ * - Only supports padding with a constant value in \b padding_value.
+ * - The input tensor and output tensor must meet the following requirements:
+ *   - The number of dimensions of \b input is no more than 8.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:3086]:
```suggestion
 *   - The number of dimensions of \b input must be less than or equal to 8.
```

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -3047,6 +3047,91 @@ mluOpAddN(mluOpHandle_t handle,
...
+ *   - The shape of the integer array of \b paddings is \p [n, 2]. Length of the first dimension of \b paddings is
+ *   \p n, which equals to the number of dimensions of \b input. Length of the second dimension of \b paddings is 2,
+ *   which means the two padding directions of before and after directions for each dimension of \b input. For each
+ *   dimension \p k, \p paddings[k, 0] and \p paddings[k, 1] specify the padding size before and after the input data
+ *   respectively.
+ *   - \p output[k] = \p paddings[k, 0] + \p input[k] + \p paddings[k, 1]. \p k represents each dimension of \b input.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:3092]: 建议生成文档看下效果，目测这个区域会有问题。。。 可以换成这种格式试试

```suggestion
 *   - ``output[k] =  paddings[k, 0] +  input[k] +  paddings[k, 1]``. \p k represents each dimension of \b input.
```

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -3047,6 +3047,91 @@ mluOpAddN(mluOpHandle_t handle,
...
+ *   \p output[k] represents the length of the dimension \p k in the output tensor. \p input[k] represents the length
+ *   of the dimension \p k in the input tensor. \p paddings[k, 0] and \p paddings[k, 1] represent the length to be
+ *   padded before and after the dimension \p k in the input tensor respectively.
+ *   - \p paddings[k, 0] and \p paddings[k, 1] >= \p -input[k]. \p k represents each dimension of \b input.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:3096]: 这个区域的格式同上，公式通常用 latex格式（黄区wiki有教程），或者 放到双点点\`\`xxx\`\`格式里

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 11]
```diff
@@ -416,6 +416,12 @@ mluOpNmsRotated
 -----------------------------
 计算旋转box的非极大值抑制。
 
+.. _pad:
+
+mluOpPad
+----------------------------------
+该算子功能是按照CONSTANT的填充方式对tensor的外围进行扩展，主要用在反向传播中的形状补齐。
```
Comment (AquamarineSu) [inline docs/bangc-docs/user_guide/9_operators/index.rst:423]: 这里再斟酌一下？头文件里是不是还提到crop的操作了？

File: bangc-ops/mlu_op.h (hunk 5 above)
Comment (gaotianyi544) [reply bangc-ops/mlu_op.h:3056]: 已修改

File: bangc-ops/mlu_op.h (hunk 7 above)
Comment (gaotianyi544) [reply bangc-ops/mlu_op.h:3125]: 已修改

File: bangc-ops/mlu_op.h (hunk 8 above)
Comment (gaotianyi544) [reply bangc-ops/mlu_op.h:3086]: 已修改

File: bangc-ops/mlu_op.h (hunk 6 above)
Comment (gaotianyi544) [reply bangc-ops/mlu_op.h:3066]: 已修改

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (gaotianyi544) [reply bangc-ops/mlu_op.h:3052]: 已经修改

File: bangc-ops/mlu_op.h (hunk 9 above)
Comment (gaotianyi544) [reply bangc-ops/mlu_op.h:3092]: 已修改

File: bangc-ops/mlu_op.h (hunk 10 above)
Comment (gaotianyi544) [reply bangc-ops/mlu_op.h:3096]: 已修改

File: docs/bangc-docs/user_guide/9_operators/index.rst (hunk 11 above)
Comment (gaotianyi544) [reply docs/bangc-docs/user_guide/9_operators/index.rst:423]: 已修改

**Review by AquamarineSu — APPROVED** [review]

**Review by PetrelYy — APPROVED** [review]
